File: rust/crates/corlinman-gateway/src/metrics.rs
//! Prometheus metrics facade for the gateway.
//!
//! Since S7 the actual metric handles live in [`corlinman_core::metrics`] so
//! the plugin / agent-client / vector crates can instrument their own call
//! sites without a reverse dependency on this crate. This module re-exports
//! every handle under its historical path so existing call sites compile
//! unchanged and the Prometheus metric *names* are identical.
//!
//! # Naming + labels
//!
//! | Metric                                     | Type      | Labels              |
//! |--------------------------------------------|-----------|---------------------|
//! | `corlinman_http_requests_total`            | CounterV  | `route`, `status`   |
//! | `corlinman_chat_stream_duration_seconds`   | HistogramV| `model`, `finish`   |
//! | `corlinman_plugin_execute_total`           | CounterV  | `plugin`, `status`  |
//! | `corlinman_plugin_execute_duration_seconds`| HistogramV| `plugin`            |
//! | `corlinman_backoff_retries_total`          | CounterV  | `reason`            |
//! | `corlinman_agent_grpc_inflight`            | IntGauge  | —                   |
//! | `corlinman_channels_rate_limited_total`    | CounterV  | `channel`, `reason` |
//! | `corlinman_vector_query_duration_seconds`  | HistogramV| `stage`             |

pub use corlinman_core::metrics::{
    encode, AGENT_GRPC_INFLIGHT, AGENT_MUTES_TOTAL, APPROVALS_TOTAL, BACKOFF_RETRIES,
    CHANNELS_RATE_LIMITED, CHAT_STREAM_DURATION, EVOLUTION_CHUNKS_DELETED, EVOLUTION_CHUNKS_MERGED,
    EVOLUTION_PROPOSALS_APPLIED, EVOLUTION_PROPOSALS_DECISION, EVOLUTION_PROPOSALS_LISTED,
    EVOLUTION_SIGNALS_DROPPED, EVOLUTION_SIGNALS_OBSERVED, EVOLUTION_SIGNALS_QUEUE_DEPTH,
    FILE_FETCHER_BYTES_TOTAL, FILE_FETCHER_FETCHES_TOTAL, HOOK_EMITS_TOTAL,
    HOOK_SUBSCRIBERS_CURRENT, HTTP_REQUESTS, LOG_FILES_REMOVED, PLUGIN_EXECUTE_DURATION,
    PLUGIN_EXECUTE_TOTAL, PROTOCOL_DISPATCH_ERRORS, PROTOCOL_DISPATCH_TOTAL,
    RATE_LIMIT_TRIGGERS_TOTAL, REGISTRY, SKILL_INVOCATIONS_TOTAL, TELEGRAM_MEDIA_TOTAL,
    TELEGRAM_UPDATES_TOTAL, VECTOR_QUERY_DURATION, WSTOOL_INVOKES_TOTAL, WSTOOL_INVOKE_DURATION,
    WSTOOL_RUNNERS_CONNECTED,
};

/// Eagerly touch every metric so names appear in `/metrics` even before the
/// first data point. Also pre-registers a `startup`-labelled sentinel so
/// label-aware series exist from the first scrape.
pub fn init() {
    corlinman_core::metrics::init();

    HTTP_REQUESTS
        .with_label_values(&["startup", "0"])
        .inc_by(0.0);
    PLUGIN_EXECUTE_TOTAL
        .with_label_values(&["startup", "ok"])
        .inc_by(0.0);
    BACKOFF_RETRIES.with_label_values(&["startup"]).inc_by(0.0);
    CHANNELS_RATE_LIMITED
        .with_label_values(&["startup", "startup"])
        .inc_by(0.0);
    CHAT_STREAM_DURATION
        .with_label_values(&["startup", "startup"])
        .observe(0.0);
    PLUGIN_EXECUTE_DURATION
        .with_label_values(&["startup"])
        .observe(0.0);
    VECTOR_QUERY_DURATION
        .with_label_values(&["startup"])
        .observe(0.0);

    // B1–B4 sentinels so each new family has at least one labelled sample.
    PROTOCOL_DISPATCH_TOTAL
        .with_label_values(&["startup"])
        .inc_by(0.0);
    PROTOCOL_DISPATCH_ERRORS
        .with_label_values(&["startup", "startup"])
        .inc_by(0.0);
    WSTOOL_INVOKES_TOTAL
        .with_label_values(&["startup", "true"])
        .inc_by(0.0);
    WSTOOL_INVOKE_DURATION
        .with_label_values(&["startup"])
        .observe(0.0);
    FILE_FETCHER_FETCHES_TOTAL
        .with_label_values(&["startup", "true"])
        .inc_by(0.0);
    FILE_FETCHER_BYTES_TOTAL
        .with_label_values(&["startup"])
        .inc_by(0.0);
    TELEGRAM_UPDATES_TOTAL
        .with_label_values(&["startup", "startup"])
        .inc_by(0.0);
    TELEGRAM_MEDIA_TOTAL
        .with_label_values(&["startup"])
        .inc_by(0.0);
    HOOK_EMITS_TOTAL
        .with_label_values(&["startup", "startup"])
        .inc_by(0.0);
    HOOK_SUBSCRIBERS_CURRENT
        .with_label_values(&["startup"])
        .set(0);
    SKILL_INVOCATIONS_TOTAL
        .with_label_values(&["startup"])
        .inc_by(0.0);
    AGENT_MUTES_TOTAL
        .with_label_values(&["startup"])
        .inc_by(0.0);
    RATE_LIMIT_TRIGGERS_TOTAL
        .with_label_values(&["startup"])
        .inc_by(0.0);
    APPROVALS_TOTAL.with_label_values(&["startup"]).inc_by(0.0);
    EVOLUTION_SIGNALS_OBSERVED
        .with_label_values(&["startup", "info"])
        .inc_by(0.0);
    EVOLUTION_SIGNALS_DROPPED.inc_by(0.0);
    EVOLUTION_SIGNALS_QUEUE_DEPTH.set(0);
    EVOLUTION_PROPOSALS_LISTED.inc_by(0.0);
    EVOLUTION_PROPOSALS_DECISION
        .with_label_values(&["startup"])
        .inc_by(0.0);
    EVOLUTION_PROPOSALS_APPLIED
        .with_label_values(&["startup", "startup"])
        .inc_by(0.0);
    EVOLUTION_CHUNKS_MERGED.inc_by(0.0);
    EVOLUTION_CHUNKS_DELETED.inc_by(0.0);
    LOG_FILES_REMOVED
        .with_label_values(&["startup"])
        .inc_by(0.0);
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The text encoder output must contain every metric name declared here.
    #[test]
    fn encode_contains_all_metric_names() {
        init();
        HTTP_REQUESTS
            .with_label_values(&["/v1/chat/completions", "200"])
            .inc();
        CHAT_STREAM_DURATION
            .with_label_values(&["claude-sonnet-4-5", "stop"])
            .observe(0.42);
        PLUGIN_EXECUTE_TOTAL
            .with_label_values(&["FooPlugin", "ok"])
            .inc();
        PLUGIN_EXECUTE_DURATION
            .with_label_values(&["FooPlugin"])
            .observe(0.12);
        BACKOFF_RETRIES.with_label_values(&["rate_limit"]).inc();
        AGENT_GRPC_INFLIGHT.inc();
        CHANNELS_RATE_LIMITED
            .with_label_values(&["qq", "group"])
            .inc();
        VECTOR_QUERY_DURATION
            .with_label_values(&["hnsw"])
            .observe(0.008);

        let body = String::from_utf8(encode()).expect("utf8");
        for needle in [
            "corlinman_http_requests_total",
            "corlinman_chat_stream_duration_seconds",
            "corlinman_plugin_execute_total",
            "corlinman_plugin_execute_duration_seconds",
            "corlinman_backoff_retries_total",
            "corlinman_agent_grpc_inflight",
            "corlinman_channels_rate_limited_total",
            "corlinman_vector_query_duration_seconds",
        ] {
            assert!(body.contains(needle), "missing {needle} in:\n{body}");
        }
        assert!(
            body.contains(r#"route="/v1/chat/completions""#),
            "missing labelled counter sample:\n{body}"
        );
    }

    #[test]
    fn http_counter_increments() {
        init();
        let before = HTTP_REQUESTS.with_label_values(&["/health", "200"]).get();
        HTTP_REQUESTS.with_label_values(&["/health", "200"]).inc();
        let after = HTTP_REQUESTS.with_label_values(&["/health", "200"]).get();
        assert_eq!(after, before + 1.0);
    }

    #[test]
    fn inflight_gauge_tracks_current_value() {
        init();
        let before = AGENT_GRPC_INFLIGHT.get();
        AGENT_GRPC_INFLIGHT.inc();
        AGENT_GRPC_INFLIGHT.inc();
        AGENT_GRPC_INFLIGHT.dec();
        assert_eq!(AGENT_GRPC_INFLIGHT.get(), before + 1);
        AGENT_GRPC_INFLIGHT.dec();
    }
}

File: python/packages/corlinman-server/src/corlinman_server/gateway/core/metrics.py
"""Prometheus metrics facade for the Python gateway.

Python port of ``rust/crates/corlinman-gateway/src/metrics.rs`` (which is
itself a re-export shell over ``corlinman-core::metrics``). The names and
labels match the Rust facade byte-for-byte so a single Grafana / alerting
config covers both runtimes.

The metric handles live on a dedicated :class:`~prometheus_client.CollectorRegistry`
called :data:`REGISTRY`. We do not use the prometheus_client default
registry so that the Python ``Process`` collector (or anything else that
auto-registers globally) doesn't accidentally show up in our /metrics
output. Callers wanting that should mount a separate scrape endpoint.

| Metric                                     | Type      | Labels              |
|--------------------------------------------|-----------|---------------------|
| ``corlinman_http_requests_total``          | Counter   | ``route``, ``status`` |
| ``corlinman_chat_stream_duration_seconds`` | Histogram | ``model``, ``finish`` |
| ``corlinman_plugin_execute_total``         | Counter   | ``plugin``, ``status``|
| ``corlinman_plugin_execute_duration_seconds`` | Histogram | ``plugin`` |
| ``corlinman_backoff_retries_total``        | Counter   | ``reason`` |
| ``corlinman_agent_grpc_inflight``          | Gauge     | — |
| ``corlinman_channels_rate_limited_total``  | Counter   | ``channel``, ``reason`` |
| ``corlinman_vector_query_duration_seconds``| Histogram | ``stage`` |
| ``corlinman_approvals_total``              | Counter   | ``decision`` |
| ``corlinman_log_files_removed_total``      | Counter   | ``reason`` |

A best-effort :func:`init` pre-touches every label set with a sentinel
``startup`` value so the names appear in ``/metrics`` from the first
scrape even if no traffic has flowed yet.
"""

from __future__ import annotations

from prometheus_client import (
    CollectorRegistry,
    Counter,
    Gauge,
    Histogram,
    generate_latest,
)

# Dedicated registry so we don't fight the global default registry.
REGISTRY: CollectorRegistry = CollectorRegistry(auto_describe=True)

# ---- HTTP layer -------------------------------------------------------------
HTTP_REQUESTS: Counter = Counter(
    "corlinman_http_requests_total",
    "HTTP requests served, labelled by route + status code",
    labelnames=("route", "status"),
    registry=REGISTRY,
)

# ---- Chat hot path ----------------------------------------------------------
CHAT_STREAM_DURATION: Histogram = Histogram(
    "corlinman_chat_stream_duration_seconds",
    "End-to-end chat stream duration in seconds",
    labelnames=("model", "finish"),
    registry=REGISTRY,
)

# ---- Plugin runtime ---------------------------------------------------------
PLUGIN_EXECUTE_TOTAL: Counter = Counter(
    "corlinman_plugin_execute_total",
    "Plugin tool invocations, labelled by plugin name + outcome",
    labelnames=("plugin", "status"),
    registry=REGISTRY,
)
PLUGIN_EXECUTE_DURATION: Histogram = Histogram(
    "corlinman_plugin_execute_duration_seconds",
    "Plugin tool invocation duration in seconds",
    labelnames=("plugin",),
    registry=REGISTRY,
)

# ---- Backoff / rate limiting -----------------------------------------------
BACKOFF_RETRIES: Counter = Counter(
    "corlinman_backoff_retries_total",
    "Number of retries the backoff helper has scheduled, by reason",
    labelnames=("reason",),
    registry=REGISTRY,
)
CHANNELS_RATE_LIMITED: Counter = Counter(
    "corlinman_channels_rate_limited_total",
    "Inbound messages dropped by channel rate-limit, by (channel, reason)",
    labelnames=("channel", "reason"),
    registry=REGISTRY,
)

# ---- gRPC inflight / vector store ------------------------------------------
AGENT_GRPC_INFLIGHT: Gauge = Gauge(
    "corlinman_agent_grpc_inflight",
    "Number of in-flight gRPC calls to the agent server",
    registry=REGISTRY,
)
VECTOR_QUERY_DURATION: Histogram = Histogram(
    "corlinman_vector_query_duration_seconds",
    "Per-stage duration of a vector store query",
    labelnames=("stage",),
    registry=REGISTRY,
)

# ---- Approvals / log retention --------------------------------------------
APPROVALS_TOTAL: Counter = Counter(
    "corlinman_approvals_total",
    "Tool approval decisions, labelled by terminal decision",
    labelnames=("decision",),
    registry=REGISTRY,
)
LOG_FILES_REMOVED: Counter = Counter(
    "corlinman_log_files_removed_total",
    "Rotated log files unlinked by the retention task, by reason",
    labelnames=("reason",),
    registry=REGISTRY,
)


def init() -> None:
    """Pre-touch every label set with a sentinel ``startup`` value so
    each metric family shows up in ``/metrics`` from the very first
    scrape. Idempotent — re-incrementing a counter by 0 is a no-op.
    """

    HTTP_REQUESTS.labels(route="startup", status="0").inc(0)
    PLUGIN_EXECUTE_TOTAL.labels(plugin="startup", status="ok").inc(0)
    BACKOFF_RETRIES.labels(reason="startup").inc(0)
    CHANNELS_RATE_LIMITED.labels(channel="startup", reason="startup").inc(0)
    CHAT_STREAM_DURATION.labels(model="startup", finish="startup").observe(0.0)
    PLUGIN_EXECUTE_DURATION.labels(plugin="startup").observe(0.0)
    VECTOR_QUERY_DURATION.labels(stage="startup").observe(0.0)
    APPROVALS_TOTAL.labels(decision="startup").inc(0)
    LOG_FILES_REMOVED.labels(reason="startup").inc(0)
    # AGENT_GRPC_INFLIGHT is unlabelled — touching .set(0) is enough.
    AGENT_GRPC_INFLIGHT.set(0)


def encode() -> bytes:
    """Render the current registry to the Prometheus text exposition
    format. Mirrors the Rust ``encode()`` shape used by ``/metrics``."""

    return generate_latest(REGISTRY)


__all__ = [
    "AGENT_GRPC_INFLIGHT",
    "APPROVALS_TOTAL",
    "BACKOFF_RETRIES",
    "CHANNELS_RATE_LIMITED",
    "CHAT_STREAM_DURATION",
    "HTTP_REQUESTS",
    "LOG_FILES_REMOVED",
    "PLUGIN_EXECUTE_DURATION",
    "PLUGIN_EXECUTE_TOTAL",
    "REGISTRY",
    "VECTOR_QUERY_DURATION",
    "encode",
    "init",
]
